File: app/src/main/java/sgtmelon/scriptum/app/control/MenuControlAnim.java
package sgtmelon.scriptum.app.control;

import android.content.Context;
import android.graphics.drawable.AnimatedVectorDrawable;
import android.os.Build;

import androidx.annotation.RequiresApi;
import sgtmelon.iconanim.office.hdlr.AnimHdlr;
import sgtmelon.scriptum.R;
import sgtmelon.scriptum.office.utils.ColorUtils;

/**
 * Класс для контроля меню с использованием анимации
 * Для версий API >= 21
 */
@RequiresApi(Build.VERSION_CODES.LOLLIPOP)
public final class MenuControlAnim extends MenuControl {

    private AnimHdlr animHdlr;

    public MenuControlAnim(Context context) {
        super(context);
    }

    @Override
    public void setupDrawable() {
        super.setupDrawable();

        final AnimatedVectorDrawable cancelOnAnim = (AnimatedVectorDrawable) ColorUtils.getDrawable(
                context, R.drawable.ic_cancel_on_anim, R.attr.clIcon
        );
        final AnimatedVectorDrawable cancelOffAnim = (AnimatedVectorDrawable) ColorUtils.getDrawable(
                context, R.drawable.ic_cancel_off_anim, R.attr.clIcon
        );

        animHdlr = new AnimHdlr(context, cancelOnAnim, cancelOffAnim);
        animHdlr.setAnimation(this);
    }

    @Override
    public void setDrawable(boolean drawableOn, boolean needAnim) {
        if (!needAnim) {
            if (drawableOn) toolbar.setNavigationIcon(cancelOn);
            else toolbar.setNavigationIcon(cancelOff);
        } else {
            animHdlr.setAnimState(drawableOn);
            if (drawableOn) {
                toolbar.setNavigationIcon(animHdlr.getAnimOn());
                animHdlr.startAnimOn();
            } else {
                toolbar.setNavigationIcon(animHdlr.getAnimOff());
                animHdlr.startAnimOff();
            }
            animHdlr.waitAnimationEnd();
        }
    }

}

File: app/src/main/java/sgtmelon/scriptum/app/control/MenuControlAnim.kt
package sgtmelon.scriptum.app.control

import android.content.Context
import android.graphics.drawable.AnimatedVectorDrawable
import android.os.Build
import android.view.View
import android.view.Window

import androidx.annotation.RequiresApi
import androidx.appcompat.widget.Toolbar
import sgtmelon.iconanim.office.hdlr.AnimHandler
import sgtmelon.scriptum.R
import sgtmelon.scriptum.office.utils.ColorUtils

/**
 * Класс для контроля меню с использованием анимации | Для версий API >= 21
 */
@RequiresApi(Build.VERSION_CODES.LOLLIPOP)
class MenuControlAnim(context: Context,
                      window: Window,
                      toolbar: Toolbar,
                      indicator: View
) : MenuControl(context, window, toolbar, indicator) {

    private val cancelOnAnim = ColorUtils.getDrawable(
            context, R.drawable.anim_cancel_enter, R.attr.clContent
    ) as AnimatedVectorDrawable?

    private val cancelOffAnim = ColorUtils.getDrawable(
            context, R.drawable.anim_cancel_exit, R.attr.clContent
    ) as AnimatedVectorDrawable?

    private val animHandler: AnimHandler = AnimHandler(context, cancelOnAnim!!, cancelOffAnim!!)

    init {
        animHandler.setAnimation(this)
    }

    override fun setDrawable(drawableOn: Boolean, needAnim: Boolean) {
        if (!needAnim) {
            toolbar.navigationIcon = if (drawableOn) cancelOn
            else cancelOff
        } else {
            animHandler.setAnimState(drawableOn)
            if (drawableOn) {
                toolbar.navigationIcon = animHandler.animOn
                animHandler.startAnimOn()
            } else {
                toolbar.navigationIcon = animHandler.animOff
                animHandler.startAnimOff()
            }
            animHandler.waitAnimationEnd()
        }
    }

}